File: unity/Runtime/Core/ScopeGuard.cs
using System;

namespace EE {
    public sealed class ScopeGuard : IDisposable {
        private bool _disposed;
        private readonly Action _undoFunction;

        public ScopeGuard(Action undoFunction) {
            _undoFunction = undoFunction;
        }

        public void Dispose() {
            if (!_disposed) {
                _undoFunction();
            }
            _disposed = true;
        }
    }
}

File: src/ts/src/core/ScopeGuard.ts
export class ScopeGuard {
    private _disposed = false;
    private readonly _undoFunction: () => void;

    public constructor(undoFunction: () => void) {
        this._undoFunction = undoFunction;
    }

    public dispose(): void {
        if (!this._disposed) {
            this._undoFunction();
        }
        this._disposed = true;
    }
}